theano/compile/function_module.py

```diff
@@ -496,6 +500,83 @@ def __copy__(self):
+            share_memory -- { boolean } Default is False. When True, two
+            function share intermediate storages(storages except input and
+            output storages). Otherwise two function will only share partial
+            storages( see method __copy__() ) and same maker. If two functions
+            share memory and allow_gc=False, this will increase executing speed
+            and save memory.
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            maker = self.maker
+            # copy Ins, so that they have different storage as their value
+            ins = copy.deepcopy(maker.inputs)
+
+            # copy fgraph and get memo
+            fg_cpy, memo = maker.fgraph.clone_get_equiv(attach_feature=False)
+
+            # construct new storage_map that map new variable to old storage
+            # so that the ensuing function shares storage with the original one
+            new_storage_map = {}
+            storage_map = self.fn.storage_map
+
+            # TODO: We could share the output storage, but we must make sure
+            # 2 different function call won't override each other values. This
+            # is already done elsewhere, so to reuse it the user would need to
+            # use Out(var, borrow=True) and maybe the mutable=True flag too.
+            # But to be safe for now as it isn't documented and we aren't sure
+            # it is well tested, we don't share the part of the storage_map.
+            for key in storage_map.keys():
+                if key not in maker.fgraph.outputs:
+                    new_storage_map[memo[key]] = storage_map[key]
+
+            input_storage = []
+            for in_ori, in_cpy in zip(maker.inputs, ins):
+                # Since we reuse original Out instances, the copied In
+                # instances should use the original variabls as their variables
+                # and updates. Otherwise the compilation will fail at function
+                # FunctionMaker._check_unused_inputs()
+                in_cpy.variable = in_ori.variable
+                in_cpy.update = in_ori.update
+
+                # share input storages if it's immutable
+                is_const = isinstance(in_ori.variable, theano.tensor.Constant)
+                if is_const or not in_ori.mutable:
+                    storage = getattr(in_ori, 'value', None)
+                    in_cpy.value = in_ori.value
+                else:
+                    storage = getattr(in_cpy, 'value', None)
+                input_storage.append(storage)
+
+            # pop out input_storage in storage_map and only use storage of In
+            # instances to initialize the make, so that we can avoid 
+            # storage conflictions in link.map_storage()
+            assert len(fg_cpy.inputs) == len(input_storage)
+            for in_var in fg_cpy.inputs:
```

Comment:
To simplify the code a little, can you merge this loop with the loop in line 542